patina_dxe_core/src/pi_dispatcher/fv.rs

```diff
@@ -0,0 +1,1702 @@
+            .files()
+            .filter(|f| {
+                f.is_err()
+                    || file_type == ffs::file::raw::r#type::ALL
+                    || f.as_ref().is_ok_and(|f| f.file_type_raw() == file_type)
+            })
+            .nth(key);
+
+        let file = match file_candidate {
+            Some(Err(err)) => return Err(err.into()),
+            Some(Ok(file)) => file,
+            _ => return Err(EfiError::NotFound),
+        };
+
+        let attributes = if (fv_attributes & fvb::attributes::raw::fvb2::MEMORY_MAPPED)
+            == fvb::attributes::raw::fvb2::MEMORY_MAPPED
+        {
+            file.fv_attributes() | fv::file::raw::attribute::MEMORY_MAPPED
+        } else {
+            file.fv_attributes()
+        };
+
+        Ok((file.name(), attributes, file.data().len(), file.file_type_raw()))
+    }
+
+    fn new_fvb_protocol(parent_handle: Option<efi::Handle>) -> Box<pi::protocols::firmware_volume_block::Protocol> {
+        Box::new(pi::protocols::firmware_volume_block::Protocol {
+            get_attributes: Self::fvb_get_attributes_efiapi,
+            set_attributes: Self::fvb_set_attributes_efiapi,
+            get_physical_address: Self::fvb_get_physical_address_efiapi,
+            get_block_size: Self::fvb_get_block_size_efiapi,
+            read: Self::fvb_read_efiapi,
+            write: Self::fvb_write_efiapi,
+            erase_blocks: Self::fvb_erase_blocks_efiapi,
+            parent_handle: parent_handle.unwrap_or(core::ptr::null_mut()),
+        })
+    }
+
+    fn new_fv_protocol(parent_handle: Option<efi::Handle>) -> Box<pi::protocols::firmware_volume::Protocol> {
+        Box::from(pi::protocols::firmware_volume::Protocol {
+            get_volume_attributes: Self::fv_get_volume_attributes_efiapi,
+            set_volume_attributes: Self::fv_set_volume_attributes_efiapi,
+            read_file: Self::fv_read_file_efiapi,
+            read_section: Self::fv_read_section_efiapi,
+            write_file: Self::fv_write_file_efiapi,
+            get_next_file: Self::fv_get_next_file_efiapi,
+            key_size: size_of::<usize>() as u32,
+            parent_handle: parent_handle.unwrap_or(core::ptr::null_mut()),
+            get_info: Self::fv_get_info_efiapi,
+            set_info: Self::fv_set_info_efiapi,
+        })
+    }
+
+    /// A helper function to generate a firmware volume block protocol instance and install it on the provided handle.
+    fn install_fvb_protocol(
+        &mut self,
+        handle: Option<efi::Handle>,
+        parent_handle: Option<efi::Handle>,
+        base_address: u64,
+    ) -> Result<efi::Handle, EfiError> {
+        let mut protocol = Self::new_fvb_protocol(parent_handle);
+
+        let protocol_ptr = protocol.as_mut() as *mut pi::protocols::firmware_volume_block::Protocol as *mut c_void;
+
+        let metadata = Metadata::new_fvb(protocol, base_address);
+
+        // save the protocol structure we're about to install in the private data.
```

Comment:
since we have an invariant on that the fv/fvb protocols should never be removed once installed, we could consider adding a usage (e.g. via `open_protocol` or a new API on ProtocolDb) to ensure that attempting to uninstall them returns ACCESS_DENIED.